moveos/moveos-stdlib/moveos-stdlib/sources/object.move

```diff
@@ -141,20 +141,73 @@ module moveos_std::object {
     /// Transfer ownership of `obj` to `recipient`. `obj` must have the
     /// `key` attribute, which (in turn) ensures that `obj` has a globally
     /// unique ID.
-    public fun transfer<T: key>(obj: T, recipient: address) {
+    public fun transfer<T: store>(obj: Object<T>, recipient: address) {
         // TODO: emit event
         transfer_internal(obj, recipient)
     }
 
     /// Freeze `obj`. After freezing `obj` becomes immutable and can no
     /// longer be transferred or mutated.
-    public native fun freeze_object<T: key>(obj: T);
+    //public fun freeze_object<T: store>(obj: Object<T>);
 
     /// Turn the given object into a mutable shared object that everyone
     /// can access and mutate. This is irreversible, i.e. once an object
     /// is shared, it will stay shared forever.
-    public native fun share_object<T: key>(obj: T);
+    //public native fun share_object<T: store>(obj: Object<T>);
+
+    native fun transfer_internal<T: store>(obj: Object<T>, recipient: address);
+
+    // Box style object
+     struct Object<T: store> {
+        id: UID,
+        value: T,
+    }
+
+    struct ObjectRef< phantom T: store> has store, drop{
+        id:  ID,
+    }
+
+    public fun new<T: store>(ctx: &mut TxContext, value: T): Object<T>{
+        let id = UID {
+            id: ID { bytes: tx_context::new_object(ctx) },
+        };
+        Object<T>{id, value}
+    }
+
+    public fun as_object_ref<T: store>(this: &Object<T>): ObjectRef<T>{
+        ObjectRef<T>{
+            id: this.id.id,
+        }
+    }
+
+    public fun borrow<T: store>(this: &Object<T>): &T{
+        &this.value
+    }
+
+    public fun borrow_mut<T: store>(this: &mut Object<T>): &mut T{
+        &mut this.value
+    }
+
+    public fun borrow_by_ref<T: store>(this: &ObjectRef<T>): &T{
+        &borrow_object(this.id).value
+    }
+
+    public fun borrow_mut_by_ref<T: store>(this: &mut ObjectRef<T>): &mut T{
```

Comment:
How to get &mut ObjectRef<T>? Is there any security problem to modify T through ObjectRef reference?

Reply:
Yes, I also think about this. Maybe we need to put a field `mutable` to ObjectRef? or define two types of reference: `ObjectRef` and `ObjectMutRef`.

The `ObjectMutRef` can be created by the mut Object.